File: backend/src/main/java/com/clickcart/service/ProviderDashboardService.java
package com.clickcart.service;

import com.clickcart.dto.provider.BookingPerformanceDto;
import com.clickcart.dto.provider.PerformanceOverviewDto;
import com.clickcart.dto.provider.PerformanceSummaryDto;
import com.clickcart.dto.provider.ProviderDashboardResponse;
import com.clickcart.dto.provider.ProviderProfileDto;
import com.clickcart.dto.provider.RecentActivityDto;
import com.clickcart.dto.provider.ScheduleItemDto;
import com.clickcart.dto.provider.ServicePerformanceDto;
import org.springframework.stereotype.Service;

import java.util.*;

@Service
public class ProviderDashboardService {

    public ProviderDashboardResponse getDashboardData(String providerId, String range, String metric) {
        ProviderProfileDto profile = new ProviderProfileDto(
                providerId != null ? providerId : "PROV-1002",
                "Kamal Perera",
                "Service Provider",
                "/assets/avatar-kamal.jpg",
                true,
                true,
                "Good morning, Kamal",
                "Here's how your business is performing today.",
                92
        );

        PerformanceSummaryDto summary = new PerformanceSummaryDto(
                new PerformanceSummaryDto.EarningMetric(125680.0, "LKR 125,680", 12.0, "up", "vs last month"),
                new PerformanceSummaryDto.JobMetric(24, 8.0, "up", "vs last month"),
                new PerformanceSummaryDto.RatingMetric(4.9, 5.0, 120),
                new PerformanceSummaryDto.CompletionRateMetric(96.0, 4.0, "up", "vs last month")
        );

        Map<String, Map<String, PerformanceOverviewDto.MetricSeries>> chartData = buildChartData();
        PerformanceOverviewDto overview = new PerformanceOverviewDto(
                "You're performing better than last month",
                chartData
        );

        BookingPerformanceDto bookingPerformance = new BookingPerformanceDto(
                5,
                8,
                24,
                1,
                38,
                new BookingPerformanceDto.StatusPercentages(12.0, 20.0, 64.0, 4.0)
        );

        List<RecentActivityDto> activity = Arrays.asList(
                new RecentActivityDto("ACT-01", "booking_accepted", "Booking Accepted", "AC Repair & Service", "#BK-0047", "2 hours ago", "calendar"),
                new RecentActivityDto("ACT-02", "job_completed", "Job Completed", "Home Cleaning", "#BK-0043", "4 hours ago", "check"),
                new RecentActivityDto("ACT-03", "payment_received", "Payment Received", "LKR 6,800", "#BK-0041", "6 hours ago", "payment"),
                new RecentActivityDto("ACT-04", "review_received", "Review Received", "5 stars • Home Cleaning", "#BK-0039", "8 hours ago", "star")
        );

        List<ServicePerformanceDto> servicePerformance = Arrays.asList(
                new ServicePerformanceDto("SVC-01", 1, "AC Repair & Service", 12, 48600.0, "LKR 48,600", 4.9, 48, "/assets/service-ac-repair.jpg"),
                new ServicePerformanceDto("SVC-02", 2, "Home Cleaning", 8, 32400.0, "LKR 32,400", 4.8, 32, "/assets/service-cleaning.jpg"),
                new ServicePerformanceDto("SVC-03", 3, "Plumbing", 6, 24680.0, "LKR 24,680", 4.7, 24, "/assets/service-plumbing.jpg")
        );

        List<ScheduleItemDto> schedule = Arrays.asList(
                new ScheduleItemDto("SCH-01", "09:00 AM", "AC Repair & Service", "Nimal Perera", "Gampaha", "Upcoming"),
                new ScheduleItemDto("SCH-02", "02:00 PM", "Home Cleaning", "Saman Kumara", "Kandy", "Upcoming"),
                new ScheduleItemDto("SCH-03", "05:30 PM", "Plumbing", "Kasun Silva", "Kurunegala", "Upcoming")
        );

        return new ProviderDashboardResponse(
                profile,
                summary,
                overview,
                bookingPerformance,
                activity,
                servicePerformance,
                schedule
        );
    }

    private Map<String, Map<String, PerformanceOverviewDto.MetricSeries>> buildChartData() {
        Map<String, Map<String, PerformanceOverviewDto.MetricSeries>> ranges = new LinkedHashMap<>();

        Map<String, PerformanceOverviewDto.MetricSeries> sevenDays = new LinkedHashMap<>();
        sevenDays.put("earnings", new PerformanceOverviewDto.MetricSeries(
                Arrays.asList("LKR 25,000", "LKR 20,000", "LKR 15,000", "LKR 10,000", "LKR 5,000", "LKR 0"),
                25000.0,
                Arrays.asList("Sep 1", "Sep 2", "Sep 3", "Sep 4", "Sep 5", "Sep 6", "Sep 7"),
                Arrays.asList(6200.0, 10400.0, 11200.0, 13800.0, 16900.0, 21400.0, 18600.0),
                "LKR "
        ));
        sevenDays.put("jobs", new PerformanceOverviewDto.MetricSeries(
                Arrays.asList("10", "8", "6", "4", "2", "0"),
                10.0,
                Arrays.asList("Sep 1", "Sep 2", "Sep 3", "Sep 4", "Sep 5", "Sep 6", "Sep 7"),
                Arrays.asList(2.0, 3.0, 4.0, 3.0, 5.0, 6.0, 4.0),
                " jobs"
        ));
        sevenDays.put("rating", new PerformanceOverviewDto.MetricSeries(
                Arrays.asList("5.0", "4.0", "3.0", "2.0", "1.0", "0.0"),
                5.0,
                Arrays.asList("Sep 1", "Sep 2", "Sep 3", "Sep 4", "Sep 5", "Sep 6", "Sep 7"),
                Arrays.asList(4.7, 4.8, 4.9, 4.8, 4.9, 5.0, 4.9),
                " / 5"
        ));
        ranges.put("7d", sevenDays);

        Map<String, PerformanceOverviewDto.MetricSeries> thirtyDays = new LinkedHashMap<>();
        thirtyDays.put("earnings", new PerformanceOverviewDto.MetricSeries(
                Arrays.asList("LKR 50,000", "LKR 40,000", "LKR 30,000", "LKR 20,000", "LKR 10,000", "LKR 0"),
                50000.0,
                Arrays.asList("Week 1", "Week 2", "Week 3", "Week 4"),
                Arrays.asList(26500.0, 31200.0, 38400.0, 29580.0),
                "LKR "
        ));
        thirtyDays.put("jobs", new PerformanceOverviewDto.MetricSeries(
                Arrays.asList("15", "12", "9", "6", "3", "0"),
                15.0,
                Arrays.asList("Week 1", "Week 2", "Week 3", "Week 4"),
                Arrays.asList(5.0, 6.0, 8.0, 5.0),
                " jobs"
        ));
        thirtyDays.put("rating", new PerformanceOverviewDto.MetricSeries(
                Arrays.asList("5.0", "4.0", "3.0", "2.0", "1.0", "0.0"),
                5.0,
                Arrays.asList("Week 1", "Week 2", "Week 3", "Week 4"),
                Arrays.asList(4.8, 4.9, 4.9, 5.0),
                " / 5"
        ));
        ranges.put("30d", thirtyDays);

        Map<String, PerformanceOverviewDto.MetricSeries> sixMonths = new LinkedHashMap<>();
        sixMonths.put("earnings", new PerformanceOverviewDto.MetricSeries(
                Arrays.asList("LKR 150,000", "LKR 120,000", "LKR 90,000", "LKR 60,000", "LKR 30,000", "LKR 0"),
                150000.0,
                Arrays.asList("Apr", "May", "Jun", "Jul", "Aug", "Sep"),
                Arrays.asList(74000.0, 86500.0, 99200.0, 112000.0, 118400.0, 125680.0),
                "LKR "
        ));
        sixMonths.put("jobs", new PerformanceOverviewDto.MetricSeries(
                Arrays.asList("30", "24", "18", "12", "6", "0"),
                30.0,
                Arrays.asList("Apr", "May", "Jun", "Jul", "Aug", "Sep"),
                Arrays.asList(14.0, 16.0, 19.0, 21.0, 22.0, 24.0),
                " jobs"
        ));
        sixMonths.put("rating", new PerformanceOverviewDto.MetricSeries(
                Arrays.asList("5.0", "4.0", "3.0", "2.0", "1.0", "0.0"),
                5.0,
                Arrays.asList("Apr", "May", "Jun", "Jul", "Aug", "Sep"),
                Arrays.asList(4.6, 4.7, 4.8, 4.8, 4.9, 4.9),
                " / 5"
        ));
        ranges.put("6m", sixMonths);

        return ranges;
    }
}

File: frontend/src/services/providerDashboardService.js
import { API_BASE_URL } from "../config/api";
import kamalAvatar from "../assets/avatar-kamal.jpg";
import acRepairImg from "../assets/service-ac-repair.jpg";
import cleaningImg from "../assets/service-cleaning.jpg";
import plumbingImg from "../assets/service-plumbing.jpg";

export const MOCK_DASHBOARD_DATA = {
  provider: {
    id: "PROV-1002",
    name: "Kamal Perera",
    role: "Service Provider",
    avatar: kamalAvatar,
    verified: true,
    online: true,
    greeting: "Good morning, Kamal",
    subtitle: "Here's how your business is performing today.",
    profileCompletion: 92,
  },
  performanceSummary: {
    totalEarnings: {
      amount: 125680,
      formatted: "LKR 125,680",
      changePercent: 12,
      trend: "up",
      periodLabel: "vs last month",
    },
    jobsCompleted: {
      count: 24,
      changePercent: 8,
      trend: "up",
      periodLabel: "vs last month",
    },
    averageRating: {
      score: 4.9,
      maxScore: 5,
      reviewCount: 120,
    },
    completionRate: {
      rate: 96,
      changePercent: 4,
      trend: "up",
      periodLabel: "vs last month",
    },
  },
  performanceOverview: {
    note: "You're performing better than last month",
    chartData: {
      "7d": {
        earnings: {
          yAxis: ["LKR 25,000", "LKR 20,000", "LKR 15,000", "LKR 10,000", "LKR 5,000", "LKR 0"],
          max: 25000,
          labels: ["Sep 1", "Sep 2", "Sep 3", "Sep 4", "Sep 5", "Sep 6", "Sep 7"],
          values: [6200, 10400, 11200, 13800, 16900, 21400, 18600],
          unit: "LKR ",
        },
        jobs: {
          yAxis: ["10", "8", "6", "4", "2", "0"],
          max: 10,
          labels: ["Sep 1", "Sep 2", "Sep 3", "Sep 4", "Sep 5", "Sep 6", "Sep 7"],
          values: [2, 3, 4, 3, 5, 6, 4],
          unit: " jobs",
        },
        rating: {
          yAxis: ["5.0", "4.0", "3.0", "2.0", "1.0", "0.0"],
          max: 5,
          labels: ["Sep 1", "Sep 2", "Sep 3", "Sep 4", "Sep 5", "Sep 6", "Sep 7"],
          values: [4.7, 4.8, 4.9, 4.8, 4.9, 5.0, 4.9],
          unit: " / 5",
        },
      },
      "30d": {
        earnings: {
          yAxis: ["LKR 50,000", "LKR 40,000", "LKR 30,000", "LKR 20,000", "LKR 10,000", "LKR 0"],
          max: 50000,
          labels: ["Week 1", "Week 2", "Week 3", "Week 4"],
          values: [26500, 31200, 38400, 29580],
          unit: "LKR ",
        },
        jobs: {
          yAxis: ["15", "12", "9", "6", "3", "0"],
          max: 15,
          labels: ["Week 1", "Week 2", "Week 3", "Week 4"],
          values: [5, 6, 8, 5],
          unit: " jobs",
        },
        rating: {
          yAxis: ["5.0", "4.0", "3.0", "2.0", "1.0", "0.0"],
          max: 5,
          labels: ["Week 1", "Week 2", "Week 3", "Week 4"],
          values: [4.8, 4.9, 4.9, 5.0],
          unit: " / 5",
        },
      },
      "6m": {
        earnings: {
          yAxis: ["LKR 150,000", "LKR 120,000", "LKR 90,000", "LKR 60,000", "LKR 30,000", "LKR 0"],
          max: 150000,
          labels: ["Apr", "May", "Jun", "Jul", "Aug", "Sep"],
          values: [74000, 86500, 99200, 112000, 118400, 125680],
          unit: "LKR ",
        },
        jobs: {
          yAxis: ["30", "24", "18", "12", "6", "0"],
          max: 30,
          labels: ["Apr", "May", "Jun", "Jul", "Aug", "Sep"],
          values: [14, 16, 19, 21, 22, 24],
          unit: " jobs",
        },
        rating: {
          yAxis: ["5.0", "4.0", "3.0", "2.0", "1.0", "0.0"],
          max: 5,
          labels: ["Apr", "May", "Jun", "Jul", "Aug", "Sep"],
          values: [4.6, 4.7, 4.8, 4.8, 4.9, 4.9],
          unit: " / 5",
        },
      },
    },
  },
  bookingPerformance: {
    pending: 5,
    confirmed: 8,
    completed: 24,
    cancelled: 1,
    total: 38,
    percentages: {
      pending: 12,
      confirmed: 20,
      completed: 64,
      cancelled: 4,
    },
  },
  recentActivity: [
    {
      id: "ACT-01",
      type: "booking_accepted",
      title: "Booking Accepted",
      service: "AC Repair & Service",
      code: "#BK-0047",
      timeAgo: "2 hours ago",
      iconType: "calendar",
    },
    {
      id: "ACT-02",
      type: "job_completed",
      title: "Job Completed",
      service: "Home Cleaning",
      code: "#BK-0043",
      timeAgo: "4 hours ago",
      iconType: "check",
    },
    {
      id: "ACT-03",
      type: "payment_received",
      title: "Payment Received",
      service: "LKR 6,800",
      code: "#BK-0041",
      timeAgo: "6 hours ago",
      iconType: "payment",
    },
    {
      id: "ACT-04",
      type: "review_received",
      title: "Review Received",
      service: "5 stars • Home Cleaning",
      code: "#BK-0039",
      timeAgo: "8 hours ago",
      iconType: "star",
    },
  ],
  servicePerformance: [
    {
      id: "SVC-01",
      rank: 1,
      title: "AC Repair & Service",
      bookings: 12,
      revenue: 48600,
      revenueFormatted: "LKR 48,600",
      rating: 4.9,
      reviewsCount: 48,
      image: acRepairImg,
    },
    {
      id: "SVC-02",
      rank: 2,
      title: "Home Cleaning",
      bookings: 8,
      revenue: 32400,
      revenueFormatted: "LKR 32,400",
      rating: 4.8,
      reviewsCount: 32,
      image: cleaningImg,
    },
    {
      id: "SVC-03",
      rank: 3,
      title: "Plumbing",
      bookings: 6,
      revenue: 24680,
      revenueFormatted: "LKR 24,680",
      rating: 4.7,
      reviewsCount: 24,
      image: plumbingImg,
    },
  ],
  todaySchedule: [
    {
      id: "SCH-01",
      time: "09:00 AM",
      service: "AC Repair & Service",
      customer: "Nimal Perera",
      location: "Gampaha",
      status: "Upcoming",
    },
    {
      id: "SCH-02",
      time: "02:00 PM",
      service: "Home Cleaning",
      customer: "Saman Kumara",
      location: "Kandy",
      status: "Upcoming",
    },
    {
      id: "SCH-03",
      time: "05:30 PM",
      service: "Plumbing",
      customer: "Kasun Silva",
      location: "Kurunegala",
      status: "Upcoming",
    },
  ],
};

const serviceImageMap = {
  "SVC-01": acRepairImg,
  "SVC-02": cleaningImg,
  "SVC-03": plumbingImg,
};

export async function fetchProviderDashboard() {
  try {
    const response = await fetch(`${API_BASE_URL}/provider/dashboard`, {
      headers: {
        Accept: "application/json",
      },
    });
    if (!response.ok) {
      console.warn("Backend API responded with non-200. Falling back to local data.", response.status);
      return MOCK_DASHBOARD_DATA;
    }
    const data = await response.json();
    const servicePerformance = (data.servicePerformance || MOCK_DASHBOARD_DATA.servicePerformance).map(
      (item) => ({
        ...item,
        image: serviceImageMap[item.id] || item.image || acRepairImg,
      })
    );
    return {
      ...MOCK_DASHBOARD_DATA,
      ...data,
      provider: {
        ...MOCK_DASHBOARD_DATA.provider,
        ...data.provider,
        avatar: kamalAvatar,
      },
      servicePerformance,
      performanceSummary: { ...MOCK_DASHBOARD_DATA.performanceSummary, ...data.performanceSummary },
      bookingPerformance: { ...MOCK_DASHBOARD_DATA.bookingPerformance, ...data.bookingPerformance },
    };
  } catch (error) {
    console.info("Using local provider dashboard dataset (API offline):", error.message);
    return MOCK_DASHBOARD_DATA;
  }
}
